bluetooth/src/commonMain/kotlin/scanner/ScanningState.kt

```diff
@@ -64,154 +64,220 @@ sealed class ScanningState : KalugaState {
+    sealed class Inactive : ScanningState(), SpecialFlowValue.NotImportant
 
+    class NotInitialized : Inactive() {
+
+        fun startInitializing(
+            scanner: BaseScanner
+        ): suspend () -> ScanningState {
+            return if (!scanner.isSupported) {
+                { NoHardware }
+            } else {
+                { Initializing(nothingDiscovered, scanner) }
+            }
+        }
+    }
+
+    data class Deinitialized(val previouslyDiscovered: Discovered, val scanner: BaseScanner) :
+        Inactive() {
+        val reinitialize: suspend () -> ScanningState =
+            { Initializing(previouslyDiscovered, scanner) }
+    }
+
+    sealed class Active :
+        ScanningState(),
+        HandleBeforeOldStateIsRemoved<ScanningState>,
+        HandleAfterNewStateIsSet<ScanningState> {
         override suspend fun beforeOldStateIsRemoved(oldState: ScanningState) {
-            if (oldState is NotInitialized) {
-                // Start monitoring Permissions and Bluetooth
-                // Stop will be called from deinitChangeState
-                scanner.startMonitoringPermissions()
-                scanner.startMonitoringSensors()
+            when (oldState) {
+                is Inactive -> {
+                    scanner.startMonitoringPermissions()
+                }
+                is Active, NoHardware -> {}
+            }
+        }
+
+        override suspend fun afterNewStateIsSet(newState: ScanningState) {
+            when (newState) {
+                is Inactive -> {
+                    scanner.stopMonitoringPermissions()
+                }
+                is Active, NoHardware -> {}
             }
         }
 
-        val revokePermission = suspend { NoBluetooth.MissingPermissions(scanner) }
+        protected abstract val scanner: BaseScanner
+        abstract val discovered: Discovered
+        val deinitialize: suspend () -> ScanningState = { Deinitialized(discovered, scanner) }
+    }
 
-        sealed class Enabled(
-            val discovered: Discovered,
-            protected val scanner: BaseScanner
-        ) : Initialized(scanner) {
+    interface Permitted : HandleBeforeOldStateIsRemoved<ScanningState>, HandleAfterNewStateIsSet<ScanningState> {
+        val revokePermission: suspend () -> NoBluetooth.MissingPermissions
+    }
 
-            val disable = suspend {
-                NoBluetooth.Disabled(scanner)
+    class PermittedHandler(val previouslyDiscovered: Discovered, val scanner: BaseScanner) : Permitted {
+
+        override val revokePermission = suspend { NoBluetooth.MissingPermissions(scanner) }
+
+        override suspend fun afterNewStateIsSet(newState: ScanningState) {
+            when (newState) {
+                is Inactive,
+                is Initializing,
+                is NoHardware,
+                is NoBluetooth.MissingPermissions -> scanner.stopMonitoringSensors()
+                else -> {}
```

Comment:
just had this in another PR, best to exhaust these.
However we were also talking about refactoring states in the performance PR, so could also leave it till then.

Reply:
Doing as part of BT rewrite